Support for multiple links to a single vehicle

acfloria opened the pull request:
- Track for each link separately if it is receiving a heartbeat from a specific vehicle
- The vehicle then selects the priority link based on all active link with the following priorities
        1. USB link
        2. Non high latency links
        3. High latency links
- Any time the active property of a link for the vehicle changes, the priority link is reevaluated.
- Any time the priority link is changed from a high latency link to a regular link and vice versa a MAV_CMD_CONTROL_HIGH_LATENCY command is sent to the vehicle.
- Add a dropdown menu to manually select the priority link from all available links to the vehicle.
- Use the link name instead of the port name for SerialLinks which is also the case for all other link types.

DonLakeFlyer commented:
I'm wondering if we can get away with not having an "Advanced Link Settings" setting. Why not just show LinkIndicator if there are multiple links and one of them is a high latency link? Removes some amount of settings clutter. Doesn't allow folks to set priority link if they don't have high latency, but not sure of the use of that and whether it would just confuse people as to what to do with it.

DonLakeFlyer commented:
Can you rebase to master so I can test. There is a conflict now.

DonLakeFlyer commented:
@acfloria Rebase?

acfloria commented:
Yes, I will work on the rebase and your suggestions tomorrow. I was busy with other projects up to today

acfloria commented on src/Vehicle/Vehicle.cc:
```diff
@@ -1604,7 +1597,12 @@ void Vehicle::_linkInactiveOrDeleted(LinkInterface* link)
+        _priorityLink.clear();
+    }
+
+    _updatePriorityLink(true, true);
```
True, thanks for pointing it out. Changed it.

acfloria commented on src/comm/LinkInterface.h:
```diff
@@ -248,10 +255,24 @@ private slots:
+     *
+     * Allocate the timer if it does not exist yet and start it.
+     */
+    void timerStart(int vehicle_id);
```
Ok, changed it to `startHeartbeatTimer` and `stopHeartbeatTimer`.

acfloria commented on src/Vehicle/Vehicle.cc:
```diff
@@ -3264,12 +3360,30 @@ void Vehicle::_vehicleParamLoaded(bool ready)
+                sendMavCommand(defaultComponentId(),
+                               MAV_CMD_CONTROL_HIGH_LATENCY,
+                               true,
+                               1.0f); // request start transmitting over high latency telemetry
```
Changed

acfloria commented on src/Vehicle/Vehicle.cc:
```diff
@@ -2357,43 +2437,59 @@ void Vehicle::setConnectionLostEnabled(bool connectionLostEnabled)
+            // communication to priority link regained
+            _connectionLost = false;
+            emit connectionLostChanged(false);
+            qgcApp()->showMessage((tr("%1 communication to priority link %2 regained")).arg(_vehicleIdSpeech()).arg(link->getName()));
```
Check added.

acfloria commented on src/Vehicle/Vehicle.cc:
```diff
@@ -2057,6 +2089,54 @@ void Vehicle::setFlightMode(const QString& flightMode)
+    }
+
+    if (newPriorityLink) {
+        _priorityLinkCommanded = true;
```
I do not see a solution here as the value of `_priorityLinkCommanded` needs to be known over time and not only when `_updatePriorityLink(...)` is called.

The reason I implemented this is the following: If you switch to HL telemetry by command and you start flying away from the ground station you might loose and regain the regular telemetry thus sending a lot of high latency activate/deactivate commands over both links. By using this flag if it is commanded nothing happens if low latency telemetry is lost/regained.

acfloria commented on src/comm/LinkInterface.h:
```diff
@@ -273,9 +294,10 @@ private slots:
     bool _enableRateCollection;
     bool _decodedFirstMavlinkPacket;    ///< true: link has correctly decoded it's first mavlink packet
+
+    QList<HeartbeatTimer*> _heartbeatTimerList;
```
Sure, I changed it. Thanks for the hint.

acfloria commented on src/ui/toolbar/LinkIndicator.qml:
```diff
@@ -0,0 +1,76 @@
+        Connections {
+            target:                 QGroundControl.multiVehicleManager
+            onActiveVehicleChanged: priorityLinkSelector.updatelinkSelectionMenu()
+        }
```
I tried to change it but was not able to make it work. How exactly would you do it?

acfloria commented:
So rebased and addressed some comments. What is currently missing:

- Removing the advanced link settings and check instead if multiple links and at least one high latency link exist

Another question I had is should we automatically switch to the high latency link if the vehicle is not armed and low latency telemetry is lost? What is your opinion on that topic?

DonLakeFlyer commented:
> Another question I had is should we automatically switch to the high latency link if the vehicle is not armed and low latency telemetry is lost?

That would seem to make sense. Although a strange case.

acfloria commented:
I removed the advanced link settings as requested.
Now the link indicator is shown if at least two links are connected to the vehicle and at least one of them is a high latency one.
In addition, the menu items of inactive links are disabled which results in a light grey text and they cannot be clicked on anymore.

acfloria commented:
Yes good to go from my side (tested multiple times on ground and in flight).

To test that part you don't need the hardware for the high latency link. You could for example connect a regular telemetry link and a usb connection with the high latency property set to true to a pixhawk to test it.
